### PR TITLE
support user-defined type annotations for struct properties

### PR DESCRIPTION
**Pull request by capfredf**
@bennn @samth

### CODE REVIEW

**Comment by AlexKnauth** [comment]
For the place in `parse-type` where you recognize `Self` and `Imp`, finding them without binding as in `(equal? (syntax-e #'id) 'Self)` worries me.

Samth Already asked
> Are these still being checked for the appropriate binding?

And it seems for your current implementation of `parse-type` the answer is no... unless I'm misunderstanding.

Should we create a binding for `Self` and `Imp` using `define-syntax`, and use `free-identifier=?` to find them with binding? Perhaps take a look at [`typed-racket/base-env/base-types-extra.rkt`](https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/base-env/base-types-extra.rkt) to see how that file does this.

The current code using `equal?` `syntax-e` that worries me is below:

https://github.com/racket/typed-racket/pull/854/files#diff-4547b05f75824d1f651616243997aa97R1070-R1074

      [id:identifier
       (define v (syntax-e #'id))
       (cond
         ;; if it's Self or Imp, we make sure it is inside the struct property
         [(equal? v 'Self)
          ....]
         [(equal? v 'Imp)
          ....]
         ....)]

**Comment by capfredf** [comment]
@AlexKnauth @samth Thank you for the review. 

Yesterday, I tried adding the literal binding for self and imp as well as the corresponding cases in `parse-type`, but it didn't work as I expected.  The cases I tried looks like:

```
[(:Self^) ...]
[(:Imp^) ...] 
```

Therefore I used the `hacky` solution. 

This morning when I was reading @AlexKnauth 's comment, it occurred to me that the patterns I had tried in `parse-type` was incorrect.

File: typed-racket-test/succeed/user-defined-sp.rkt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#lang typed/racket
+
+(: prop:foo (StructProperty (-> Self Number)))
+(define-values (prop:foo _1 _2) (make-struct-type-property 'foo))
```
[the remark this answers is not in the document]
Comment (AlexKnauth) [reply typed-racket-test/succeed/user-defined-sp.rkt]: `_1` would be the `foo?` predicate, and according to the type of `make-struct-type-property` at [typed-racket/base-env/base-env.rkt:1234](https://github.com/racket/typed-racket/pull/854/files#diff-ef79b6b89ee78c4b6ada735b8e32a899R1234) it has type `(-> Any Boolean)`.

Currently there are no types that describe "an instance of any struct that implements this property" so far, so it couldn't have a more descriptive type like `(-> Any Boolean : (Struct-Has-Property prop:foo))`. So is `(-> Any Boolean)` good enough for now?

[the remark this answers is not in the document]
Comment (bennn) [reply typed-racket-test/succeed/user-defined-sp.rkt]: It'd be good to put types for `_1` and `_2` in the test

File: typed-racket-lib/typed-racket/private/parse-type.rkt [hunk 2]
```diff
@@ -55,6 +56,11 @@
 ;; current-term-names when parsing ψ
 (define current-local-term-ids (make-parameter '()))
 
+;; current-in-struct-prop: Paremeter<Boolean>
+;; This parameter is used to assure that some types
+;; are only allowed to used in struct properties
```
Comment (bennn) [inline typed-racket-lib/typed-racket/private/parse-type.rkt]: This comment is not very clear. I get the idea, but please fix the grammar, use Racket syntax for the type, and maybe mention Self & Imp

**Comment by samth** [comment]
Yes, @bennn makes a good point about the name `Struct-Property`. That's also in keeping with the rest of TR's names.

**Comment by samth** [comment]
Two remaining things:

1. `Struct-Property` should be contravariant.
2. Make sure that the docs are complete and up to date.

**Comment by capfredf** [comment]
@samth I have squashed the previous commits and reworded the message.

**Comment by bennn** [comment]
Make sure to update the scribble docs. They need to: (1) say `Struct-Property` and (2) talk about user-defined type annotations.

**Pull request merged**